apps/aecore/lib/aecore/chain/worker.ex

```diff
@@ -161,8 +161,13 @@ defmodule Aecore.Chain.Worker do
     Logger.info(fn ->
       "Added block ##{new_block.header.height} with hash #{Base.encode16(new_block_hash)}, total tokens: #{inspect(total_tokens)}"
     end)
+
     ## Store new block to disk
     Persistence.write_block_by_hash(new_block)
+
+    ## Store new chain_state to disk
+    Persistence.write_chain_state_by_pubkey(updated_chain_states)
```

Comment:
it would be more efficient to update the changed chainstates (calculated earlier) in the database, instead of a full rewrite everytime

Reply:
right now you write the full chainstate for each block to disk, we should find a way to only write the changes affected by this block

Reply:
We would have to probably use some special database with some kind of versioning.

Reply:
we don't need versioning for our blockchain implementation. Additional indexes for better APIs should be implemented in external programms

Reply:
We could use SQL probably and save changes as new rows, but this will be probably slow unless configured properly and configuring can be impossible.

Reply:
We can also Implement our own versioning by generating saving only differences. Then to read data back one would have to apply the changes since the first block.

Reply:
hmm...

Reply:
It would be very slow and inefficient, the plan is to use rocksdb for the current state and think about versioning later.

Reply:
So I think there are two separate problems:
1. We are storing all states (for all blocks) in memory currently. This can be solved in future by just deleting data in memory.
2. When we want to store information in rox we have to write the whole state not just changes.

Reply:
So the question to @meivantodorov is: How can we split the chain_state for rox so only the updates are written to disk.

Reply:
Currently we have:
 - all blocks in memory
 - the latest chainstate in memory
 - the blocks downloaded by sync in memory

Ideas to persist or swap:
 - save everything instead of last 10/100 (difficulty adjustment range) as `hash -> block` to rocksdb
 - save each account state `account -> state` to rocksdb and update if this account changes in one block
 - download all blocks for sync to disk, load them from 1 to top to validate